internal/consumer/dlq.go

```diff
@@ -36,97 +38,148 @@ type (
-		producer:    producer,
-		retryPolicy: newDLQRetryPolicy(),
-		tally:       scope,
-		logger:      logger.With(zap.String("topic", topic), zap.String("cluster", cluster)),
+		stopC     chan struct{}
+		doneC   chan struct{}
+
+		scope     tally.Scope
+		logger    *zap.Logger
+		lifecycle *util.RunLifecycle
 	}
-}
+)
 
-// Add blocks until successfully enqueuing the given
-// message into the error queue
-func (d *dlqImpl) Add(m kafka.Message) error {
-	sm, err := d.newSaramaMessage(m)
+// NewDLQ returns a DLQ writer for writing to a specific DLQ topic
+func NewDLQ(topic, cluster string, client sarama.Client, scope tally.Scope, logger *zap.Logger) (DLQ, error) {
+	asyncProducer, err := sarama.NewAsyncProducerFromClient(client)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	return backoff.Retry(func() error { return d.add(sm) }, d.retryPolicy, d.isRetryable)
+	return newBufferedDLQ(topic, cluster, asyncProducer, scope, logger), nil
 }
 
-func (d *dlqImpl) Close() {
-	d.producer.Close()
-}
-
-func (d *dlqImpl) add(m *sarama.ProducerMessage) error {
-	_, _, err := d.producer.SendMessage(m)
-	if err != nil {
-		d.logger.Error("error sending message to DLQ", zap.Error(err))
+func newBufferedDLQ(topic, cluster string, producer sarama.AsyncProducer, scope tally.Scope, logger *zap.Logger) *bufferedErrorTopic {
+	return &bufferedErrorTopic{
+		topic: topic,
+		cluster: cluster,
+		producer: producer,
+		stopC: make(chan struct{}),
+		doneC: make(chan struct{}),
+		scope: scope,
+		logger: logger,
+		lifecycle: util.NewRunLifecycle(fmt.Sprintf("dlqProducer-%s-%s", topic, cluster), logger),
 	}
-	return err
 }
 
-func (d *dlqImpl) isRetryable(err error) bool {
-	return true
+// NewDLQNoop returns a DLQ that drops everything on the floor
+// this is used only when DLQ is not configured for a consumer
+func NewDLQNoop() DLQ {
+	return &dlqNoop{}
 }
+func (d *dlqNoop) Add(m kafka.Message) error { return nil }
+func (d *dlqNoop) Stop()                     {}
+func (d *dlqNoop) Start() error              { return nil }
 
-func newDLQRetryPolicy() backoff.RetryPolicy {
-	policy := backoff.NewExponentialRetryPolicy(dlqRetryInitialInterval)
-	policy.SetMaximumInterval(dlqRetryMaxInterval)
-	policy.SetExpirationInterval(backoff.NoInterval)
-	return policy
+func (d *bufferedErrorTopic) Start() error {
+	return d.lifecycle.Start(func() error {
+		go d.asyncProducerResponseLoop()
+		return nil
+	})
 }
 
-func (d *dlqImpl) newSaramaMessage(m kafka.Message) (*sarama.ProducerMessage, error) {
-	key, err := d.encodeDLQMetadata(m)
-	if err != nil {
-		return nil, err
-	}
-	return &sarama.ProducerMessage{
-		Topic: d.topic,
-		Key:   sarama.ByteEncoder(key),
-		Value: sarama.ByteEncoder(m.Value()),
-	}, nil
+func (d *bufferedErrorTopic) Stop() {
+	d.lifecycle.Stop(func() {
+		close(d.stopC)
+		<-d.doneC
+	})
 }
 
-func (d *dlqImpl) encodeDLQMetadata(m kafka.Message) ([]byte, error) {
+func (d *bufferedErrorTopic) Add(m kafka.Message) error {
```

Comment:
add tally metrics on here on the number of messages added / succeeded / failed, if this is not already tracked by the top level consumer (which uses this DLQ)

Reply:
I'll put metrics in another pass. Need to align metrics names and tags with what we currently have internally. I've put a TODO to mark this.